android/KMEA/app/src/main/java/com/tavultesoft/kmea/LanguagesSettingsActivity.java

```diff
@@ -143,54 +144,62 @@ public void onBackPressed() {
+      super(context, RESOURCE, storage, new AdapterFilter<Dataset.LanguageDataset, Dataset, Void>() {
+        public List<Dataset.LanguageDataset> selectFrom(Dataset dataset, Void dummy) {
+          // Filter out any languages without installed keyboards.  This can occur with ad-hoc
+          // lexical model installations.
+          ArrayList<Dataset.LanguageDataset> languages = new ArrayList<>();
+
+          for(Dataset.LanguageDataset language: dataset.asList()) {
+            if(language.keyboards.size() > 0) {
+              languages.add(language);
+            }
+          }
+
+          return languages;
         }
+      }, null);
+    }
+
+    @Override
+    public View getView(int position, View convertView, ViewGroup parent) {
+      Dataset.LanguageDataset data = this.getItem(position);
+      ViewHolder holder;
+
+      // If we're being told to reuse an existing view, do that.  It's automatic optimization.
+      if (convertView == null) {
+        convertView = LayoutInflater.from(getContext()).inflate(RESOURCE, parent, false);
+        holder = new ViewHolder();
+
+        holder.img = convertView.findViewById(R.id.image1);
+        holder.textLang = convertView.findViewById(R.id.text1);
+        holder.textCount = convertView.findViewById(R.id.text2);
+        convertView.setTag(holder);
+      } else {
+        holder = (ViewHolder) convertView.getTag();
       }
 
-      // Increment an existing language count
-      if (languageIndex != -1) {
-        HashMap<String, String> matchingLanguageInfoInfo = list.get(languageIndex);
-        String keyboardCount = matchingLanguageInfoInfo.get(KMManager.KMKey_KeyboardCount);
-        if (keyboardCount.equalsIgnoreCase(oneKeyboard)) {
-          matchingLanguageInfoInfo.put(KMManager.KMKey_KeyboardCount, twoKeyboards);
-        } else {
-          int index = keyboardCount.indexOf(" ");
-          Integer currentCount = Integer.parseInt(keyboardCount.substring(1, index));
-          matchingLanguageInfoInfo.put(KMManager.KMKey_KeyboardCount, String.format("(%d keyboards)", currentCount + 1));
-        }
-        list.set(languageIndex, matchingLanguageInfoInfo);
+      holder.textLang.setText(data.name);
+      holder.img.setImageResource(R.drawable.ic_arrow_forward);
+
+      if(data.keyboards.size() == 1) {
+        holder.textCount.setText("(1 keyboard)");
       } else {
-        // Otherwise, add new entry of language name and count of 1
-        HashMap<String, String> lgInfo = new HashMap<String, String>();
-        lgInfo.put(KMManager.KMKey_LanguageID, languageID);
-        lgInfo.put(KMManager.KMKey_LanguageName, languageName);
-        lgInfo.put(KMManager.KMKey_KeyboardCount, oneKeyboard);
-        lgInfo.put(KMManager.KMKey_Icon, String.valueOf(R.drawable.ic_arrow_forward));
-        lgInfo.put("isEnabled", "true");
-        list.add(lgInfo);
+        holder.textCount.setText("(" + data.keyboards.size() + " keyboards)");
```

Comment:
This is of course hard to localize ... :) We can cross that bridge later but we will need to refactor it!